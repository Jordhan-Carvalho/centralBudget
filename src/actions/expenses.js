import db from '../firebase/firebase';


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
  });

    
  export const startAddExpense = (expenseData = {}) => {
    // only works bc redux middleware (redux thunk)
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0,
      } = expenseData;
      const expense = {description, note, amount, createdAt};
      //save the data to firebase (aynsc call) (used the return to testing porpouse)
        const ref = await db.ref(`users/${uid}/expenses`).push(expense);
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    };
  };
  
  // REMOVE_EXPENSE
  export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
  });

  export const startRemoveExpense = (expense) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
        const id = expense.id;
  try{
  await db.ref(`users/${uid}/expenses/${id}`).remove();
        dispatch(removeExpense(expense));
      } catch (error) {
        alert(error);
      }
      };
      };
  
  // EDIT_EXPENSE
  export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
  });

  export const startEditExpense = (id, updates) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      try{
        await db.ref(`users/${uid}/expenses/${id}`).update(updates);
        dispatch(editExpense(id, updates));
      } catch (error) {
        console.log(error);
      }
    };
  };

  // SET_EXPENSES
  export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
  });

  // export const startSetExpenses
  export const startSetExpenses = () => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
  // 1. fetch all expense data once
  //2. Parse the data to array
  const expenses = []; 
  await db.ref(`users/${uid}/expenses`).once('value', (snapshot) => {
        snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
        });
        // 3. dispatch SET_EXPENSES
        dispatch(setExpenses(expenses));
      };
      };