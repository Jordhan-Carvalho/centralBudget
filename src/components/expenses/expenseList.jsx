import React from "react";
import { connect } from "react-redux";
import ListItem from "../common/listItem";
import selectVisible from "../../selectors/visibleTrans";

export const ExpenseList = props => {
  return (
    <div>
      {props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        <ListItem type={"expense"} expenses={props.expenses} />
      )}
    </div>
  );
};

// redux call
const mapStateToProps = state => {
  return {
    expenses: selectVisible(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
