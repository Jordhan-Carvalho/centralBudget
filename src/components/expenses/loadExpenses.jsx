import React, { Component } from "react";
import { Link } from "react-router-dom";
import ExpenseList from "./expenseList";
import ListFilters from "../common/listFilters";
import TransSumary from "../common/transSumary";

class LoadExpenses extends Component {
  state = {};
  render() {
    return (
      <React.Fragment>
        <h1>Expense List</h1>
        <hr />
        <ListFilters />
        <TransSumary type={"expense"} />
        <hr />
        <ExpenseList />
        <hr />
        <Link to="/create">
          <button className="btn btn-dark">Add Expense</button>
        </Link>
      </React.Fragment>
    );
  }
}

export default LoadExpenses;
