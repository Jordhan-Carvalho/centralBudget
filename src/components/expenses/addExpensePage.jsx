import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "../common/form";
import { startAddExpense } from "../../actions/expenses";

export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push("/expenses");
  };

  render() {
    return (
      <React.Fragment>
        <h1>Add Expense</h1>
        <Form onSubmit={this.onSubmit} />
      </React.Fragment>
    );
  }
}

// to simplify props.dispatch(addExpense(expense)) to props.onSubmit(expense)
const mapDispatchToProps = dispatch => {
  return {
    startAddExpense: expense => dispatch(startAddExpense(expense))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
