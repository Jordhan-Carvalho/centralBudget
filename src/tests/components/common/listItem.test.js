import React from 'react';
import { shallow } from 'enzyme';
import ListItem from '../../../components/common/listItem';
import expenses from '../../fixtures/expenses';

test('should render list of expenses', () => {
    const wrapper = shallow(<ListItem expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
});