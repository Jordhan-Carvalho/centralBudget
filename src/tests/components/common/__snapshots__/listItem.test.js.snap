// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render list of expenses 1`] = `
<table
  className="table"
>
  <thead
    className="thead-dark"
  >
    <tr>
      <th
        scope="col"
      >
        Title
      </th>
      <th
        scope="col"
      >
        Amount
      </th>
      <th
        scope="col"
      >
        Note
      </th>
      <th
        scope="col"
      >
        Date
      </th>
      <th
        scope="col"
      >
        Edit
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="1"
    >
      <td>
        Gum
      </td>
      <td>
        $1.95
      </td>
      <td />
      <td>
        December 31st, 1969
      </td>
      <td>
        <Link
          to="/edit/undefined/1"
        >
          Edit
        </Link>
      </td>
    </tr>
    <tr
      key="2"
    >
      <td>
        Rent
      </td>
      <td>
        $1,095.00
      </td>
      <td />
      <td>
        December 27th, 1969
      </td>
      <td>
        <Link
          to="/edit/undefined/2"
        >
          Edit
        </Link>
      </td>
    </tr>
    <tr
      key="3"
    >
      <td>
        Credit Card
      </td>
      <td>
        $45.00
      </td>
      <td />
      <td>
        January 4th, 1970
      </td>
      <td>
        <Link
          to="/edit/undefined/3"
        >
          Edit
        </Link>
      </td>
    </tr>
  </tbody>
</table>
`;
